ampligraph/latent_features/models/ConvE.py

```diff
@@ -0,0 +1,1158 @@
+
+                self.bn_vars = {'batchnorm_input': {'beta': np.zeros(shape=[emb_img_dim]),
+                                                    'gamma': np.ones(shape=[emb_img_dim]),
+                                                    'moving_mean': np.zeros(shape=[emb_img_dim]),
+                                                    'moving_variance': np.ones(shape=[emb_img_dim])},
+                                'batchnorm_conv': {'beta': np.zeros(shape=[nfilters]),
+                                                   'gamma': np.ones(shape=[nfilters]),
+                                                   'moving_mean': np.zeros(shape=[nfilters]),
+                                                   'moving_variance': np.ones(shape=[nfilters])},
+                                'batchnorm_dense': {'beta': np.zeros(shape=[1]),       # shape = [1] for batch norm
+                                                    'gamma': np.ones(shape=[1]),
+                                                    'moving_mean': np.zeros(shape=[1]),
+                                                    'moving_variance': np.ones(shape=[1])}}
+
+            if self.embedding_model_params['use_bias']:
+                self.bias = tf.get_variable('activation_bias', shape=[1, len(self.ent_to_idx)],
+                                            initializer=tf.zeros_initializer(), dtype=tf.float32)
+
+        else:
+            raise NotImplementedError('ConvE not implemented when dealing with large graphs.')
+
+    def _get_model_loss(self, dataset_iterator):
+        """Get the current loss including loss due to regularization.
+        This function must be overridden if the model uses combination of different losses(eg: VAE).
+        This function must be overridden if the model uses combination of different losses(eg: VAE).
+
+        Parameters
+        ----------
+        dataset_iterator : tf.data.Iterator
+            Dataset iterator.
+
+        Returns
+        -------
+        loss : tf.Tensor
+            The loss value that must be minimized.
+        """
+
+        # training input placeholder
+        self.x_pos_tf, self.y_true = dataset_iterator.get_next()
+
+        # list of dependent ops that need to be evaluated before computing the loss
+        dependencies = []
+
+        # if the graph is large
+        if self.dealing_with_large_graphs:
+            raise NotImplementedError('ConvE not implemented when dealing with large graphs.')
+
+        # run the dependencies
+        with tf.control_dependencies(dependencies):
+
+            # look up embeddings from input training triples
+            e_s_pos, e_p_pos, e_o_pos = self._lookup_embeddings(self.x_pos_tf)
+
+            # Get positive predictions
+            self.y_pred = self._fn(e_s_pos, e_p_pos, e_o_pos)
+
+            # Label smoothing and/or weighting is applied within Loss class
+            loss = self.loss.apply(self.y_true, self.y_pred)
+
+            if self.regularizer is not None:
```

Comment:
Should we also regularize the weights?

Reply:
Good question.

Reply:
There's no regularization in the ConvE github reference code, so I'll leave it for now.